src/contrib/cluster/Akka.Cluster.Sharding/ShardCoordinator.cs

```diff
@@ -6,572 +6,2242 @@
 //-----------------------------------------------------------------------
 
 using System;
-using Akka.Actor;
-using Akka.Event;
+using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Linq;
 using System.Threading.Tasks;
+using Akka.Actor;
+using Akka.Event;
 
 namespace Akka.Cluster.Sharding
 {
     using ShardId = String;
 
-    internal interface IShardCoordinator
+    /// <summary>
+    /// Singleton coordinator that decides where to allocate shards.
+    /// </summary>
+    internal class ShardCoordinator
```

Comment:
In the Scala code, `ShardCoordinator` is an abstract class inheriting from `ReceiveActor`.
Please re-write this class.

Reply:
it's not possible to use it same as in scala, since there are no traits in c# and we need to use the same logic in persistent and non-persistent actor. The implementation is similar to the previous implementation, except before it was implemented via common static class. Same as how  e.g. `AtLeastOnceDelivery` is implemented to workaround c# limitations

Reply:
Makes sense